pico/runtime/src/main/java/io/helidon/pico/runtime/AbstractServiceProvider.java

```diff
@@ -202,10 +202,16 @@ public Phase currentActivationPhase() {
     /**
      * Used to access the current pico services instance assigned to this service provider.
      *
+     * @param expected true to indicate that the return values is expected to be non-empty
      * @return the pico services assigned to this service provider
      */
-    public PicoServices picoServices() {
-        return Objects.requireNonNull(picoServices, description() + ": picoServices should have been previously set");
+    @Override
+    public Optional<PicoServices> picoServices(boolean expected) {
+        if (expected) {
```

Comment:
The whole idea of return an optional is to leave the decision whether it is expected or not on the caller of the method.
This method should just `return Optional.ofNullable(picoServices)`
The caller should throw an exception if they should be available

Reply:
I wanted the provider to control the exception and not the caller. I'll address it in a slightly modified way.

Reply:
pushed